Adds "Randomize on Randomizer Generation" options to Audio and Cosmetics Editors

Glought opened the pull request:
Adds "Randomize on Randomizer Generation" options to Audio and Cosmetics Editors

 * Adds "Randomize All Music and Sound Effects on Randomizer Generation" option to the Audio Editor.

 * Adds "Randomize All on Randomizer Generation" option to the Cosmetics Editor.

When enabled, these options randomize audio and/or cosmetics during a new randomizer generation.
It's functionally the same as pressing the "Randomize All"  Button.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/3427430638.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/3427434655.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/3427435207.zip)
<!--- section:artifacts:end -->

Glought committed:
Add "Randomize on Randomizer Generation" options to Audio and Cosmetics Editors

 * Added "Randomize All Music and Sound Effects on Randomizer Generation" to the Audio Editor.

 * Added "Randomize All on Randomizer Generation" to the Cosmetics Editor.

When enabled, these options randomize audio and/or cosmetics during a new randomizer generation.

Glought committed:
Added "OnRandomizerGeneration" Hook.
 * It is executed on Randomizer Generation.

Glought committed:
Changed AudioEditor and CosmeticsEditor "Randomize all on Randomizer Generation" options to use the "OnRandomizerGeneration" Hook.

Pepper0ni commented:
~~(At least) cosmetics on randomiser generation causes a settings eating crash if a seed is generated immediately after another seed stops generation (Mash the generate button) with address not mapped to object. Will investigate to see if it's specific to this code or some concurrency issue with the various configs saving.~~

Nevermind, got it without mashing, investigating further.

Pepper0ni reviewed (approved):
I'm going to tentatively approve right now, as the crashing issue seems to be the fault of randomiser code not anything in this PR, and it otherwise works as intended.

Glought commented:
> ~(At least) cosmetics on randomiser generation causes a settings eating crash if a seed is generated immediately after another seed stops generation (Mash the generate button) with address not mapped to object. Will investigate to see if it's specific to this code or some concurrency issue with the various configs saving.~
> 
> Nevermind, got it without mashing, investigating further.

Odd though my testing i experience no crashing and just tested now mashing  generating  randos in both the ingame file menu and SoH Gui menu I didn't get any crashing. I'm on Windows 10 if that helps in case you use mac or Linux.

Pepper0ni commented:
I am on linux, but I suspect it's more likely the debugger slowing down saving and making the timings different, as the likely cause is it trying to save to the json while the cosmetic editor is.

serprex commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3611,6 +3611,9 @@ bool GenerateRandomizer(std::string seed /*= ""*/) {
     if (CVarGetInteger(CVAR_GENERAL("RandoGenerating"), 0) == 0) {
         randoThread = std::thread(&GenerateRandomizerImgui, seed);
+
+        GameInteractor_ExecuteOnRandomizerGeneration();
```
I think this hook would be more interesting as an on-generation-complete

ie when the success jingle is played

This could then be used to hook code that wants to process spoiler json

Maybe it fixes Pepper0ni's race too

Glought replied:
I made made some changes based on your suggestion.

Glought committed:
Renamed "OnRandomizerGeneration" to "OnGenerationCompletion.
Renamed "GameInteractor_ExecuteOnRandomizerGeneration" to "GameInteractor_ExecuteOnGenerationCompletion"
Moved "GameInteractor_ExecuteOnGenerationCompletion" from "GenerateRandomizer" to the end of "GenerateRandomizerImgui".

Glought commented:
> I am on linux, but I suspect it's more likely the debugger slowing down saving and making the timings different, as the likely cause is it trying to save to the json while the cosmetic editor is.

I've made some changes: audio and cosmetics are now randomized after the Randomizer finishes generating. Hopefully, that resolves your crashing issue.

Pepper0ni commented:
> I've made some changes: audio and cosmetics are now randomized after the Randomizer finishes generating. Hopefully, that resolves your crashing issue.

Thanks, though i think the real fix is nothing to do with your code, I'll push a PR for it

Glought committed:
Merge branch 'develop' into Randomize-Audio-and-Cosmetics-on-Randomizer-Generation

Malkierian commented on soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.cpp:
```diff
@@ -306,6 +306,10 @@ void GameInteractor_ExecuteOnFileChooseMain(void* gameState) {
     GameInteractor::Instance->ExecuteHooks<GameInteractor::OnFileChooseMain>(gameState);
 }
 
+void GameInteractor_ExecuteOnGenerationCompletion() {
```
Because this is getting called from a C++ file, and is highly unlikely to ever be called from a C file, I feel this bridge is unnecessary, and ExecuteHooks should be called directly from `randomizer.cpp`

Glought replied:
I pushed a commit based on your suggestions. If you have any other suggestions please let me know.

Glought committed:
Removed "GameInteractor_ExecuteOnGenerationCompletion()" from "GameInteractor_Hooks.h" and "GameInteractor_Hooks.cpp"
The "OnGenerationCompletion" hook is now called directly at the end of "GenerateRandomizerImgui" in "randomizer.cpp"

Pepper0ni reviewed (approved):
Still works fine.
